### PR TITLE
[NuGet Symbol Server] Gallery upload - Refactoring

### PR DESCRIPTION
**Pull request by shishirx34**
This is a preparation PR for adding support to gallery for symbols package upload.

I have refactored code out from the `ApiController` for the symbols upload, so that we can reuse the code in `PackagesController` and keep common code in single place. Also, as part of this I had to refactor/re-write some tests for the coverage. I do not touch the Package upload flow(which should also be refactored) as part of this PR, since its a lot more refactor and out of scope for the symbol upload flow. 

Also, refactored some part of `PackagesController` to reuse upload page in gallery. In subsequent PRs I will be doing additive changes to enhance this controller for symbols upload.

Apologies for the big PR(refactoring is a pain)

### CODE REVIEW

File: src/NuGetGallery.Core/PackageMetadataExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+        private const string SymbolPackageTypeName = "SymbolsPackage";
+        private static readonly ClientPackageType SymbolPackageType = new ClientPackageType(SymbolPackageTypeName, ClientPackageType.EmptyVersion);
+
+        public static bool IsSymbolPackage(this PackageMetadata metadata)
+        {
+            var packageTypes = metadata.GetPackageTypes();
+            return packageTypes.Any()
+                && packageTypes.Count() == 1
+                && packageTypes.First() == SymbolPackageType;
+        }
```
Comment (cristinamanum) [inline src/NuGetGallery.Core/PackageMetadataExtensions.cs]: Why not packageTypes.Where.(..).Any()?

File: src/NuGetGallery/Controllers/PackagesController.cs [hunk 2]
```diff
@@ -244,155 +250,161 @@ public virtual async Task<JsonResult> UploadPackage(HttpPostedFileBase uploadFil
...
                     }
-                }
 
-                await _uploadFileService.SaveUploadFileAsync(currentUser.Key, uploadStream);
+                    return Json(HttpStatusCode.Conflict, new[] { message });
+                }
             }
 
+            await _uploadFileService.SaveUploadFileAsync(currentUser.Key, uploadStream);
+            uploadStream.Dispose();
```
Comment (skofman1) [inline src/NuGetGallery/Controllers/PackagesController.cs]: uploadStream  should be disposed by the caller of this method (UploadPackage(HttpPostedFileBase uploadFile)) as part of finally block

File: src/NuGetGallery/Helpers/ZipArchiveHelpers.cs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+                if (entryInTheFuture != null)
+                {
+                    entry = entryInTheFuture;
+                    return true;
+                }
+            }
+
+            return false;
+        }
+    }
```
Comment (cristinamanum) [inline src/NuGetGallery/Helpers/ZipArchiveHelpers.cs]: I think that this is an old method, but can we add few comments about its purpose? What is are found ore than one entries in the future?  It does not matter which is returned ?

File: src/NuGetGallery/Services/SymbolPackageValidationResult.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+
+        /// <summary>
+        /// There is no corresponding nuget package available for the uploaded symbols package.
+        /// </summary>
+        MissingPackage,
+
+        /// <summary>
+        /// There is a symbols package already present, pending validations.
+        /// </summary>
+        SymbolsPackageExists
```
Comment (cristinamanum) [inline src/NuGetGallery/Services/SymbolPackageValidationResult.cs]: Why not SymbolPackagePendingValidation ?

**Pull request merged**